New Lua main menu settings

PilzAdam opened the pull request:
A new setting tab that contains all available settings, while still being easy to use.

The settings are stored in a file with a special markup, that is parsed dynamically by the Lua main menu. It contains information about expected type, default value, and other constraints. This allows the menu to show different dialogs for different types of settings (e.g. a dropdown for enums, or a slider for ints).

What is still missing:
- ~~A `path` type; maybe we can even open a file browser in the mainmenu.~~
- A `flags` type for mapgen flags; a list of checkboxes with all available options would be cool here.
- A `noise_params` type; since they are always the same structure and they are excessively used I think they deserve a special treatment in the GUI.
- Fancy input for `float`  and `int`; if max and min are specified then use a slider, else use a input field with input validation.
- The engine supports arbitrary structs in the settings, but there is none that uses this (except noise params, but they get a special treatment); this is low priority.
- Maybe a fancier GUI.

Also the `settings.txt` file probably needs a lot of improvement. Maybe open up a forum thread after this is merged so the actual end-users can help to modify it.

Also there are functions to automatically generate `minetest.conf.example` and a translation file that can be parsed by xgettext.

PilzAdam committed:
New settings tab contain all possible settings

Settings are automatically parsed from builtin/settingtypes.txt
The edit dialog automatically adjust based on the type of setting

kilbith commented:
Preview here : https://github.com/minetest/minetest/issues/3254#issuecomment-148681373

PilzAdam commented:
Added `path` type with fancy file browse dialog.

kahrl commented:
I like it! (Finally my "tree" tablecolumn is getting some usage! :D)

Some points:
- I think the basic settings dialog should stay and the new settings dialog should be accessible through a button labeled "Advanced Settings" or something like that. Having access to all settings at once is probably overwhelming for most users (notice how e.g. Firefox has about:config, while their main settings dialog is something much less powerful, but simpler).
- Sometimes when you open a subtree by clicking on the +, the table scrolls back to the selected item, which is very annoying.
- After you edit a setting, all the subtrees are closed again.

Fixing the last two points probably requires some changes to GUIFormSpecMenu / GUITable. I'm not sure the last point can be easily fixed at all (maybe allow showing a new formspec as a modal dialog while the old formspec is kept around in the background?)

kilbith commented:
There's a bug : 
1. Open the file browser from a Edit dialog.
2. Close the browser dialog.
3. Cancel or Save the Edit dialog.

You end-up with a screwed formspec : https://lut.im/GXOtev11h9/deMPracckRIrp8xy.png

est31 commented:
Very nice!
Will give it a closer look later.

PilzAdam commented:
@kahrl The old settings tab contained arbitrary settings. I agree that the number of settings can be overwhelming, but I think that the categories should be used to fix that instead of the old settings tab.

Your last two points are general issues with formspecs and are out of scope for this PR.

@kilbith I noticed this bug in other places where the file dialog is used, too, so I think that it's unrelated to this PR.

kilbith commented:
@PilzAdam I did not switched of locale, the bug happens on the texture path dialog as well (without changing anything).

PilzAdam commented:
@kilbith maybe the file dialog changes the locale automatically. Can you verify this by querying the locale with `os.setlocale (nil, "numeric")` in a few places?

kahrl commented:
@PilzAdam I don't agree that the current settings tab is arbitrary. It focused on graphics settings, which I find to be the type of setting (aside from key bindings) I am interested the most in when configuring a game, since they really determine whether the game is playable on my computer or not.

Yeah, the current settings tab is missing some useful options (e.g. sound volume) which should be fixed.

My last two points are general issues, yes, but they make the settings tab awkward to use.

kilbith commented:
@PilzAdam Done, no influence on the bug.

ShadowNinja commented:
@PilzAdam: Irrlicht is known to reset the locale when you open a file chooser.   This was why the option to add a mod from a local file was removed from the mod store menu shortly before a release.

PilzAdam commented:
@ShadowNinja https://github.com/PilzAdam/minetest/blob/lua_settings/builtin/mainmenu/tab_settings.lua#L298 would resetting the locale here (and in the _cancel variant) be enough to work around this?

@kahrl If you are mostly interesting in graphic settings, then just open up the graphic section. I have nothing against an aggregate view by default in the settings tab, but I don't like the current one.

@kilbith log the return value of that function call.

kilbith commented:
@PilzAdam The log : http://paste.ubuntu.com/12801621/ (I use Arch Linux with the recommended settings : http://paste.ubuntu.com/12801630/)

ShadowNinja commented:
@PilzAdam: I'm not sure.  It's worth a try.

est31 commented:
Ok, I've downloaded and tried it. I really like it. Some things I think have to be fixed before merging:

* settings.txt can be misinterpreted by some people to be the actual file there is configuration in. However, when they open it and read carefully, they'll see fast that their assumption is wrong. Any other name ideas?
* If you look at a setting's value in the tree view, you see "true/false". If you klick at "Edit", you see "Enable/Disable".
* Perhaps the border on the right is too small?
* It should probably print a warning during file parsing if the description is longer than 6 lines. This makes it easy to spot too long descriptions, e.g. for `debug_log_level`.
* Perhaps should we use different notations for sub-sections? [abc][/def] can be easily mistaken with bbcode. What about `[section]`  `[*subsection]`  `[**subsubsection]`?
* I really wonder how it happens, but if a setting has no description, it displays the german po file for me. (Example `main_menu_script`)
* some dev has to read through every line of code, I haven't done that yet, only tested.

And of course the thigs kahrl has said. I think this setting dialog should be behind an "advanced settings" button on the settings tab. Most needed settings are/should be there.

PilzAdam commented:
@est31
* Maybe `defaultsettings.txt`?
* Changed; they now also display "Enabled" / "Disabled" in the overview
* Thats GUI design, so I don't care ;-) (this can be made fancy later)
* This depends on font_size and GUI scale, so not possible with formspecs; maybe change it into a table so it's scrollable?
* Changed.
* That was a bug, thanks for catching it. (I passed "" to gettext)
* I'm a dev! ;-)

Don't you think that this can be done by good grouping of the settings?

est31 commented:
`settingtypes.txt`?

est31 commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -15,497 +15,446 @@
+	local currentComment = ""
+	
+	local line = file:read("*line")
+	while line do
```
Perhaps the body of the while can be put into its own function? Then you wouldn't need the `if handled then`, and could simply do `return true` instead of `handled = true`.

est31 commented:
The [naming guidelines](http://dev.minetest.net/Lua_code_style_guidelines#Naming) for lua prefer `lowercase_underscore_style` for both functions and variables.

est31 commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -15,497 +15,446 @@
 
-	if fields["sb_gui_scaling"] then
-		local event = core.explode_scrollbar_event(fields["sb_gui_scaling"])
+	formspec = formspec:sub(1, -2) -- remove trailing comma
```
I know this is very unlikely, but what happens if `settings` is {}?

est31 commented:
:+1: except comments. The last one is very nitpicky, so its ok if its not addressed.

PilzAdam commented:
@est31 fixed all your comments.

est31 commented:
Good for squash + merge now, for me.

kahrl commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -15,497 +15,504 @@
-}
+	-- settings
+	local first_part, name, readable_name, type =
+			line:match("^(([%w%._-]+)[%s]+%(([^%)]*)%)[%s]+([%w_]+)[%s]*)")
```
- Should there be a $ at the end of the regex?
- 'type' shadows a builtin lua function.

PilzAdam replied:
* No, since the type dependent parameters follow
* Is this really an issue here?

kahrl commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -15,497 +15,504 @@
-				"textlist[8.33,2.7;4,1;;#888888" .. fgettext("Waving Leaves") .. ";0;true]" ..
-				"textlist[8.33,3.2;4,1;;#888888" .. fgettext("Waving Plants") .. ";0;true]"
+	local current_level = 0
+	for _, entry in pairs(settings) do
```
Shouldn't this be ipairs? (pairs() iterates in an unspecified order)

kahrl commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -15,497 +15,504 @@
+			"#    http://wiki.minetest.net/\n" ..
+			"\n"
+
+	for _, entry in pairs(settings) do
```
ditto ipairs

est31 replied:
ipairs is bad as it doesn't preserve order, and if you automatically create the .conf.example it will bloat up and obfuscate the diff.

kahrl commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -15,497 +15,504 @@
+			"// It conatins a bunch of fake gettext calls, to tell xgettext about the strings in config files\n" ..
+			"// To update it, refer to the bottom of builtin/mainmenu/tab_settings.lua\n\n" ..
+			"fake_function() {\n"
+	for _, entry in pairs(settings) do
```
ditto ipairs
